apps/arweave/src/ar_kv.erl

```diff
@@ -2,659 +2,548 @@
 
 -behaviour(gen_server).
 
--export([start_link/0, open/2, open/3, open/4, put/3, get/2, get_next_by_prefix/4, get_next/2,
-		get_prev/2, get_range/2, get_range/3, delete/2, delete_range/3, count/1]).
+-export([
+	start_link/0, create_ets/0, open/2, open/3, open/4, put/3, get/2,
+	get_next_by_prefix/4, get_next/2, get_prev/2, get_range/2, get_range/3,
+	delete/2, delete_range/3, count/1
+]).
 
 -export([init/1, handle_cast/2, handle_call/3, handle_info/2, terminate/2]).
 
 -include_lib("arweave/include/ar.hrl").
 -include_lib("arweave/include/ar_config.hrl").
 -include_lib("eunit/include/eunit.hrl").
 
+-define(WITH_DB(Name, Callback), with_db(Name, ?FUNCTION_NAME, Callback)).
+-define(WITH_ITERATOR(Name, IteratorOptions, Callback), with_iterator(Name, ?FUNCTION_NAME, IteratorOptions, Callback)).
+
+-define(DEFAULT_ROCKSDB_DATABASE_OPTIONS, #{
+	create_if_missing => true,
+	create_missing_column_families => true,
+
+	%% enable atomic_flush for dbs that utilize column families
+	atomic_flush => true,
+
+	%% these are default values, but they must not be overriden;
+	%% otherwise the syncWAL will not work.
+	allow_mmap_reads => false,
+	allow_mmap_writes => false
+}).
+
+-record(db, {
+	%% name may be undefined in short intervals before opening the database,
+	%% or reopening the database (which implies close and open operations).
+	%% It may happen in case of opening the database with column families.
+	%% NB: records with undefined db_handle must not be stored in the ETS table.
+	name :: term() | undefined,
+	filepath :: file:filename_all(),
+	db_options :: rocksdb:db_options(),
+	%% db_handle may be undefined in short intervals before opening the database,
+	%% or reopening the database (which implies close and open operations).
+	%% NB: records with undefined db_handle must not be stored in the ETS table.
+	db_handle :: rocksdb:db_handle() | undefined,
+
+	%% column families only fields, must be set to undefined for plain databases.
+	cf_names = undefined :: [term()],
+	cf_descriptors = undefined :: [rocksdb:cf_descriptor()],
+	cf_handle = undefined :: rocksdb:cf_handle()
+}).
+
 -record(state, {}).
 
+
+
 %%%===================================================================
 %%% Public interface.
 %%%===================================================================
 
+
+
 start_link() ->
 	gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
 
+
+
+%% @doc Creates a named ETS table.
+%% This function is used within `ar_kv_sup` as well as `ar_test_node` modules.
+create_ets() ->
+	ets:new(?MODULE, [set, public, named_table, {keypos, #db.name}]).
```

Comment:
Whoa, I didn't know you could use a record field as a `keypos` value. Perfect solution.

Reply:
Yea, the `Var#rec.field` notation is desugared into `element(Var, #rec.field)`, which is desugared into `element(Var, n)`. So the `#rec.field` can be used anywhere in the code to get the field number in the tuple, it is just an integer.

Reply:
Awesome.